client/storage-monitor/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+use std::{
+	path::{Path, PathBuf},
+	time::Duration,
+};
+
+const LOG_TARGET: &str = "storage-monitor";
+
+/// Error type used in this crate.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+	#[error("IO Error")]
+	IOError(#[from] Errno),
+	#[error("Out of storage space: available {0}MB, required {1}MB")]
+	StorageOutOfSpace(u64, u64),
+}
+
+/// Parameters used to create the storage monitor.
+#[derive(Default, Debug, Clone, Args)]
+pub struct StorageMonitorParams {
+	/// Required available space on database storage. If available space for DB storage drops below
+	/// the given threshold, node will be gracefully terminated. If `0` is given monitoring will be
+	/// disabled.
+	#[arg(long = "db-storage-threshold", value_name = "MB", default_value_t = 1000)]
+	pub threshold: u64,
+
+	/// How often available space is polled.
+	#[arg(long = "db-storage-polling-period", value_name = "SECONDS", default_value_t = 5, value_parser = clap::value_parser!(u32).range(1..))]
+	pub polling_period: u32,
+}
+
+/// Storage monitor service: checks the available space for the filesystem for fiven path.
+pub struct StorageMonitorService {
+	/// watched path
+	path: PathBuf,
+	/// number of megabytes that shall be free on the filesystem for watched path
+	threshold: u64,
+	/// storage space polling period (seconds)
+	polling_period: u32,
+}
+
+impl StorageMonitorService {
+	/// Creates new StorageMonitorService for given client config
+	pub fn try_spawn(
+		parameters: StorageMonitorParams,
+		database: DatabaseSource,
+		spawner: &impl SpawnEssentialNamed,
+	) -> Result<(), Error> {
+		Ok(match (parameters.threshold, database.path()) {
+			(0, _) => {
+				log::info!(
+					target: LOG_TARGET,
+					"StorageMonitorService: threshold `0` given, storage monitoring disabled",
+				);
+			},
+			(_, None) => {
+				log::warn!(
+					target: LOG_TARGET,
+					"StorageMonitorService: no database path to observe",
+				);
+			},
+			(threshold, Some(path)) => {
+				log::debug!(
+					target: LOG_TARGET,
+					"Initializing StorageMonitorService for db path: {:?}",
+					path,
+				);
+
+				Self::check_free_space(&path, threshold)?;
+
+				let storage_monitor_service = StorageMonitorService {
+					path: path.to_path_buf(),
+					threshold,
+					polling_period: parameters.polling_period,
+				};
+
+				spawner.spawn_essential(
+					"storage-monitor",
+					None,
+					Box::pin(storage_monitor_service.run()),
+				);
+			},
+		})
+	}
+
+	/// Main monitoring loop, intended to be spawned as essential task. Quits if free space drop
+	/// below threshold.
+	async fn run(self) {
+		loop {
+			tokio::time::sleep(Duration::from_secs(self.polling_period.into())).await;
+			if Self::check_free_space(&self.path, self.threshold).is_err() {
+				break
+			};
+		}
+	}
+
+	/// Returns free space in MB, or error if statvfs failed.
+	fn free_space(path: &Path) -> Result<u64, Error> {
+		statvfs(path)
+			.map(|stats| stats.blocks_available() * stats.block_size() / 1_000_000)
```

Comment:
I'm try to build latest Substrate, and compiling failed

```
error[E0308]: mismatched types
   --> client/storage-monitor/src/lib.rs:121:44
    |
121 |             .map(|stats| stats.blocks_available() * stats.block_size() / 1_000_000)
    |                                                     ^^^^^^^^^^^^^^^^^^ expected `u32`, found `u64`

error[E0277]: cannot multiply `u32` by `u64`
   --> client/storage-monitor/src/lib.rs:121:42
    |
121 |             .map(|stats| stats.blocks_available() * stats.block_size() / 1_000_000)
    |                                                   ^ no implementation for `u32 * u64`
    |
    = help: the trait `Mul<u64>` is not implemented for `u32`
    = help: the following other types implement trait `Mul<Rhs>`:
              <&'a u32 as Mul<num_bigint::bigint::BigInt>>
              <&'a u32 as Mul<num_bigint::biguint::BigUint>>
              <&'a u32 as Mul<u32>>
              <&'b u32 as Mul<&'a num_bigint::bigint::BigInt>>
              <&'b u32 as Mul<&'a num_bigint::biguint::BigUint>>
              <&u32 as Mul<&u32>>
              <u32 as Mul<&'a num_bigint::bigint::BigInt>>
              <u32 as Mul<&'a num_bigint::biguint::BigUint>>
            and 6 others

error[E0308]: mismatched types
   --> client/storage-monitor/src/lib.rs:120:3
    |
119 |       fn free_space(path: &Path) -> Result<u64, Error> {
    |                                     ------------------ expected `Result<u64, Error>` because of return type
120 | /         statvfs(path)
121 | |             .map(|stats| stats.blocks_available() * stats.block_size() / 1_000_000)
122 | |             .map_err(Error::from)
    | |_________________________________^ expected `u64`, found `u32`
    |
    = note: expected enum `Result<u64, _>`
               found enum `Result<u32, _>`

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
```

can reproduce by `cargo build --release -p sc-storage-monitor`

Reply:
Could you share output of `rustup show` on machine where it failed?

Reply:
Seems `[blocks_available](https://docs.rs/nix/latest/x86_64-apple-darwin/nix/sys/statvfs/struct.Statvfs.html#method.blocks_available)` return i32 for Apple/Darwin (https://docs.rs/libc/0.2.139/x86_64-apple-darwin/libc/type.fsblkcnt_t.html). We should have this in CI. Working on fix.

Reply:
fix proposal: https://github.com/paritytech/substrate/pull/13234

Reply:
Thank you for reporting this.